waggle-dance-integration-tests/src/main/java/com/hotels/bdp/waggledance/junit/ServerSocketRule.java

```diff
@@ -45,104 +38,49 @@
       throw new RuntimeException(e);
     }
   }
 
-  @Override
-  protected void before() throws Throwable {
-    listener.accept(null, new CompletionHandler<AsynchronousSocketChannel, Void>() {
-      @Override
-      public void completed(AsynchronousSocketChannel channel, Void attachment) {
-        listener.accept(null, this);
-        handle(channel);
-      }
-
-      @Override
-      public void failed(Throwable exception, Void attachment) {
-        LOG.warn("Failed to process request", exception);
-      }
-    });
-  }
-
-  private void handle(final AsynchronousSocketChannel channel) {
-    LOG.info("Submitting request");
-    requests.offer(executor.submit(new Callable<Void>() {
-      @Override
-      public Void call() throws Exception {
-        synchronized (output) {
-          try (InputStream input = Channels.newInputStream(channel)) {
-            ByteStreams.copy(input, output);
-          } catch (IOException e) {
-            throw new RuntimeException("Error processing user request", e);
-          }
-        }
-        return null;
-      }
-    }));
-  }
-
   @Override
   protected void after() {
-    LOG.info("Socket closing, got '{}' requests left", requests.size());
-    executor.shutdown();
+    LOG.info("Socket closing, handled {} requests", requests);
     try {
-      executor.awaitTermination(1L, TimeUnit.SECONDS);
-      listener.close();
-    } catch (InterruptedException e) {
-      throw new RuntimeException(e);
+      serverSocketChannel.close();
     } catch (IOException e) {
       throw new RuntimeException(e);
     }
   }
 
   public byte[] getOutput() {
-    return waitAndgetOutput(1, TimeUnit.SECONDS);
-  }
-
-  /**
-   * Waits for timeout to get any requests and then flushes every request and returns the result
-   *
-   * @param timeout
-   * @param unit
-   * @return bytes received
-   */
-  public byte[] waitAndgetOutput(long timeout, TimeUnit unit) {
     try {
-      Thread.sleep(unit.toMillis(timeout));
-    } catch (InterruptedException e) {
-      throw new RuntimeException(e);
-    }
-    awaitRequests(requests.size(), timeout, unit);
-    synchronized (output) {
-      return output.toByteArray();
-    }
-  }
-
-  private void awaitRequests(int requestCount, long timeout, TimeUnit unit) {
-    while (requestCount > 0) {
-      if (requests.peek() == null) {
-        throw new RuntimeException("No requests have been received");
-      }
-      try {
-        requests.peek().get(timeout, unit);
-        requests.poll();
-        requestCount--;
-      } catch (Exception e) {
-        throw new RuntimeException("Error while waiting for request completion", e);
+      SocketChannel socketChannel = serverSocketChannel.accept();
+      while (socketChannel != null) {
+        requests++;
+        try (InputStream input = Channels.newInputStream(socketChannel)) {
+          ByteStreams.copy(input, output);
+        }
+        socketChannel = serverSocketChannel.accept();
```

Comment:
Why do you need to do this?

Reply:
an `accept()` for every request, see the `expectMultipleRequests` test it loops through this 3 times